### ISSUE TITLE
target position is wrong for _perform_click in some cases

### ISSUE THREAD

**Issue by nixgnef** [pywinauto/pywinauto#915]
## Expected Behavior

`pywinauto.controls.hwndwrapper._perform_click`  should simulate the click behavior correcttly.

## Actual Behavior

when the  “Dispaly -> Scale and layout -> Change the size of text, apps and other items”  setting  for  windows 10 OS  is  **NOT 100%**， the target position of 
`pywinauto.controls.hwndwrapper._perform_click `
may need to be adjusted according to the scale ratio. 

## Steps to Reproduce the Problem

  1.  set  thet “dispaly ->  scale and layout -> change the size of texts apps and other items”  to be 200% 

  2.  use pywinauto to  get an item from a **SysTreeView32** control in an app，and then call the click() method  for this item 

## Short Example of Code to Demonstrate the Problem
           
```
            tree_view = app.top_window().child_window(
                control_id=129, class_name="SysTreeView32"
            )
            target_item = tree_view.wrapper_object().get_item('query')
            target_item.click()
```

## a raw method to solve this problem 

inser the following code into `pywinauto.controls.hwndwrapper._perform_click` near line 1663 in hwndwrapper.py 

```
    scale_factor = win32print.GetDeviceCaps(win32gui.GetDC(ctrl.handle), win32con.LOGPIXELSX) / 96
    if scale_factor != 1.0:
        coords = [int(item*scale_factor) for item in coords]
```
after 
```
    if absolute:
        coords = ctrl.client_to_screen(coords)
```
## postscript
`pywinauto.mouse._perform_click_input` may also need to be adjusted.

## Specifications

  - Pywinauto version:  0.6.8
  - Python version and bitness: 3.7
  - Platform and OS: windows 10

**Comment by nixgnef** [comment]
In the above case,  using  select() to substitute click() also works:  
```
            tree_view = app.top_window().child_window(
                control_id=129, class_name="SysTreeView32"
            )
            target_item = tree_view.wrapper_object().get_item(['query','target'])
            target_item.select()
```

**Comment by vasily-v-ryabov** [comment]
This is the same problem as #37 (pretty old issue).

**Comment by nixgnef** [comment]
@vasily-v-ryabov I had seen [#37 ](https://github.com/pywinauto/pywinauto/issues/37) , but I still thought it's necessary to open a new one，because I've figured out that this issue is caused by the DPI(dots per inch) setting of windows OS, which is not apparently mentioned in  [#37 ](https://github.com/pywinauto/pywinauto/issues/37) .

**Comment by airelil** [comment]
BTW, the 96 is the default DPI?

**Comment by nixgnef** [comment]
@airelil  96 is the DPI corresponding to the 100% scale ratio, however, the default scale ratio may not be 100%(in my machine is 125%).  In fact, this problem is more complicated then I had realized.  Because there are three DPI modes(see the doc from microsoft [DPI Awareness Mode](https://docs.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows#dpi-awareness-mode) , and for each mode we may need a specified way to get the right coordinates.

**Comment by airelil** [comment]
@nixgnef, thanks, that's good article. We've done some work for DPI awareness: https://github.com/pywinauto/pywinauto/blob/51e67932650702d5f3f2d1c7efa91150ac3b2a74/pywinauto/windows/win32functions.py#L752
Seems as it needs to be reviewed again. BTW, the article also mentions new added enum
```C
PROCESS_PER_MONITOR_DPI_AWARE_V2 = 3, /* Newer HiDPI type for Windows 10 1607+ */
```

**Comment by nixgnef** [comment]
@airelil  I've found addtional docs relevent to this issue:
 1)  [DWM scaling](https://docs.microsoft.com/en-us/windows/win32/learnwin32/dpi-and-device-independent-pixels#dwm-scaling) (My issue may be caused by the DWM scaling in which case the coordinates getting from the API and the display do not match) 
 2)  [High DPI Settings in Windows](http://www.kynosarges.org/WindowsDpi.html)  
Finally, people who use pywinauto include me should thank you contributors of this project. :smiley:

**Comment by airelil** [comment]
So you suggest adding `scale` parameter to user facing methods, such as: `HwndWrapper.click()`, `double_click` and `close_click`?
https://github.com/pywinauto/pywinauto/blob/1b474df126444ac45729709bb6e9f39fccf57877/pywinauto/controls/hwndwrapper.py#L731

**Comment by nixgnef** [comment]
@airelil 🤝 Yes. 
     1) This issue will not occur in the well programmed new windows applications.
     2) The comment for this newer added parameter can remind users the existence of this issue and to avoid this issue, users can  try to set the scale ratio to 100% for each monitor, or to adjust the scale ratio for each click() to match the corresponding monitor setting.

**Comment by airelil** [comment]
Yep, we can consider it. Also, I had a bit time today to check this. I noticed that calling `SetProcessDpiAwareness` with the new `PROCESS_PER_MONITOR_DPI_AWARE_V2` (3) might help. You can try to make this change in your pywinauto code too. 
However, I couldn't find the value to be documented. The [PROCESS_DPI_AWARENESS](https://docs.microsoft.com/en-au/windows/win32/api/shellscalingapi/ne-shellscalingapi-process_dpi_awareness) still has old enum definitions.  So I want to try switching to [SetProcessDpiAwarenessContext](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setprocessdpiawarenesscontext) instead, but need to find proper values for [DPI_AWARENESS_CONTEXT](https://docs.microsoft.com/en-au/windows/win32/hidpi/dpi-awareness-context) handle

**Comment by nixgnef** [comment]
@airelil  :handshake:   Ok，I will have a try at my leisure.

**Comment by laoyur** [comment]
okay... it cost me one day to figure out why _listview_item.click() chooses wrong item...  :(

**Comment by nixgnef** [comment]
> okay... it cost me one day to figure out why _listview_item.click() chooses wrong item... :(
I also spent a little time to find the real cause，so I opened  this issue  to remind other users。

**Comment by bratao** [comment]
Yeah, I have been bitten by this too. Any chance of fixing it on a upcoming version?

**Comment by nixgnef** [comment]
> Yeah, I have been bitten by this too. Any chance of fixing it on a upcoming version?

To handle all situtions could be tedious, this issue may be caused by the OS setting, the application setting and the compatibility between the OS and the application.
